src/sacn/receiver_state.c

```diff
@@ -231,24 +231,31 @@ void remove_receiver_from_thread(SacnReceiver* receiver, socket_close_behavior_t
  */
 etcpal_error_t add_receiver_sockets(SacnReceiver* receiver)
 {
-  etcpal_error_t res = kEtcPalErrOk;
+  etcpal_error_t ipv4_res = kEtcPalErrNoNetints;
+  etcpal_error_t ipv6_res = kEtcPalErrNoNetints;
 
   if (supports_ipv4(receiver->ip_supported))
   {
-    res = sacn_add_receiver_socket(receiver->thread_id, kEtcPalIpTypeV4, receiver->keys.universe,
-                                   receiver->netints.netints, receiver->netints.num_netints, &receiver->ipv4_socket);
-
-    if ((res == kEtcPalErrNoNetints) && supports_ipv6(receiver->ip_supported))
-      res = kEtcPalErrOk;  // Try IPv6.
+    ipv4_res =
+        sacn_add_receiver_socket(receiver->thread_id, kEtcPalIpTypeV4, receiver->keys.universe,
+                                 receiver->netints.netints, receiver->netints.num_netints, &receiver->ipv4_socket);
   }
 
-  if ((res == kEtcPalErrOk) && supports_ipv6(receiver->ip_supported))
+  if (((ipv4_res == kEtcPalErrOk) || (ipv4_res == kEtcPalErrNoNetints)) && supports_ipv6(receiver->ip_supported))
   {
-    res = sacn_add_receiver_socket(receiver->thread_id, kEtcPalIpTypeV6, receiver->keys.universe,
-                                   receiver->netints.netints, receiver->netints.num_netints, &receiver->ipv6_socket);
+    ipv6_res =
+        sacn_add_receiver_socket(receiver->thread_id, kEtcPalIpTypeV6, receiver->keys.universe,
+                                 receiver->netints.netints, receiver->netints.num_netints, &receiver->ipv6_socket);
   }
 
-  return res;
+  if ((ipv4_res == kEtcPalErrNoNetints) && (ipv6_res == kEtcPalErrNoNetints))
```

Comment:
Good catch on cleaning up the v4 socket if the function ultimately returns an error. I'll put that in.

First it's worth establishing that I'm treating NoNetints as encompassing an additional case where the receiver does not support IPv4 or IPv6, thus it is the default starting value of ipv4_res and ipv6_res. I figured this works because technically there are "no netints" for the corresponding version that is not supported, since those netints are not allowed to be used. This is ultimately a useful simplification.

The code you propose doesn't handle the case where v4=NoNetints and v6 != NoNetints. It ends up returning v4 (NoNetints) when it should actually return whatever is in v6.

I think the code I wrote covers this case, and in addition the case that you mentioned. If v4 = ErrOk and v6 != ErrOk/ErrNoNetints, my code will return v6 consistently. In the other case, where v4 != ErrOk/ErrNoNetints and v6 = ErrOk, it's actually not possible for v6 to be set to ErrOk because the v4 error is treated as a terminal error, so that case doesn't matter and my code is simplified accordingly.

Please let me know if you're seeing a specific error combination that is not handled correctly.

Reply:
https://github.com/ETCLabs/sACN/pull/52/commits/6a1aa053a2f5d9a7fc5b7df5f6f4cdf437d83f6f